test/test_abstraction.py

```diff
@@ -74,6 +77,85 @@ def test_http_upgrade(self, monkeypatch):
         assert c._conn._sock == 'totally a non-secure socket'
 
 
+def TestHTTPConnection_SocketLevel(SocketLevelTest):
+
+    def test_version_after_http_upgrade(self):
+        self.set_up()
+        self.secure = False
+
+        send_event = threading.Event()
+
+        def socket_handler(listener):
+            sock = listener.accept()[0]
+
+            # We should get the initial request.
+            data = b''
+            while not data.endswith(b'\r\n\r\n'):
+                data += sock.recv(65535)
+
+            send_event.wait()
+
+            # We need to send back a response.
+            resp = (
+                b'HTTP/1.1 101 Upgrade\r\n'
+                b'Server: socket-level-server\r\n'
+                b'Content-Length: 0\r\n'
+                b'Connection: upgrade\r\n'
+                b'Upgrade: h2c\r\n'
+                b'\r\n'
+            )
+            sock.send(resp)
+
+            sock.close()
+
+        self._start_server(socket_handler)
+
+        c = HTTPConnection(self.host, self.port)
+        assert c.version is HTTPVersion.http11
+        c.request('GET', '/')
+        send_event.set()
+        resp = c.get_response()
+        assert c.version is HTTPVersion.http20
+        assert resp.version is HTTPVersion.http20
+
+    def test_version_after_tls_upgrade(self):
+        self.set_up()
+        self.secure = False
+
+        send_event = threading.Event()
+
+        def socket_handler(listener):
+            sock = listener.accept()[0]
+
+            # We should get the initial request.
+            data = b''
+            while not data.endswith(b'\r\n\r\n'):
+                data += sock.recv(65535)
+
+            send_event.wait()
+
+            # We need to send back a response.
+            resp = (
+                b'HTTP/1.1 101 Switching Protocols\r\n'
+                b'Server: socket-level-server\r\n'
+                b'Content-Length: 0\r\n'
+                b'Connection: upgrade\r\n',
+                b'Upgrade: TLS/1.0, HTTP/2.0'
```

Comment:
Uh...what is this? This test doesn't make any sense at all, and frankly suggests that the upgrade code in hyper is broken. =P